docs/cudf/source/internals.md

```diff
@@ -0,0 +1,419 @@
+`Buffer.empty`:
+
+```python
+>>> buf = Buffer.empty(10)
+>>> print(buf.size)
+10
+>>> print(buf._owner)
+<rmm._lib.device_buffer.DeviceBuffer object at 0x7f6055baa890>
+```
+
+### Column and its subclasses
+
+A `Column` is a sequence of values, any number of which may be "null".
+
+```cython
+cdef class Column:
+    def __init__(self, data, size, dtype, mask=None, offset=0, children=()):
+        pass
+```
+
+* `data`: A `Buffer` containing data for the Column's elements.
+* `size`: Number of elements (**not** size in bytes) in the Column
+* `dtype`: The element type
+* `mask`: A `Buffer` representing the Column's null mask
+* `offset`: (used in slices) An integer offset into the first element of the slice
+* `children`: A tuple of "children" Columns.
+  Children columns enable complex column types such as
+  Columns with non-fixed width elements (e.g., strings).
+
+Columns are specialized based on the type of data they contain:
+
+```cython
+# Column class hierarchy
+
+cdef class Column  # extension type
+
+class ColumnBase(Column)  # Python class
+
+class NumericalColumn(ColumnBase)
+class DatetimeColumn(ColumnBase)
+class StringColumn(ColumnBase)
+class CategoricalColumn(ColumnBase)
+```
+
+Note that `Column` is an
+[extension type](https://cython.readthedocs.io/en/latest/src/userguide/extension_types.html)
+defined using `cdef class`,
+a special type that can directly interface with C/C++ code.
+
+### Table and Frame
+
+A `Table` is a collection of Columns, with an optional `index` (itself a `Table`).
+
+```cython
+cdef class Table:
+    def __init__(self, data=None, index=None):
+        pass
+```
+
+* `data`: An `OrderedDict` of Columns
+* `index`: A `Table` representing the index of this Table
+
+```cython
+# Table class hierarchy
+
+cdef class Table  # extension type
+
+class Frame(Table)  # Python class
+
+class Index(Frame)
+class Series(Frame)
+class DataFrame(Frame)
+```
+
+Like `Column`, `Table` is an extension type.
+
+## Interfacing with libcudf <a name="interfacing"></a>
+
+`libcudf` is a C++/CUDA library that provides
+GPU accelerated implementations of common data operations
+such as sort, filter and groupby.
+Most operations on cuDF objects like DataFrames or Series
+are performed by libcudf under the hood.
+
+### Why Cython
+
+cuDF uses [Cython](https://cython.org) to interface with libcudf.
+While there are other options for using C/C++ code from Python,
+we use Cython becuase:
+
+1. It is used by other libraries in the PyData ecosystem such as
+   Pandas and NumPy.
+2. It supports  *modern* C++ features
+   such as smart pointers and move semantics.
+3. Cython code is much more similar to Python than it is to C/C++,
+   making it easier to use by Python programmers.
+
+### Wrapping a libcudf API - example
+
+This is the sort of thing best explained using an example,
+so we consider wrapping the libcudf function `gather`
+and exposing it in Python:
+
+```c++
+lib.pxd
```

Comment:
should this be `something.hpp` ?

Reply:
No - it's currently `lib.pxd`. Although now I'm thinking that we likely want to have a separate `.pxd` file corresponding to each `.pyx` file. I'll add a section on code organization to make this clear.

Reply:
Sorry - it absolutely should be `copying.hpp`. Fix coming up